File: codes/ruby/chapter04/answer_A16.rb
# 入力（A, B が 0 番目から始まっていることに注意）
n = gets.to_i
a = gets.split.map(&:to_i)
b = gets.split.map(&:to_i)

# 動的計画法
dp = Array.new(n + 1)
dp[1] = 0
dp[2] = a[0]
(3..n).each do |i|
	dp[i] = [dp[i-1] + a[i-2], dp[i-2] + b[i-3]].min
end

# 出力
puts dp[n]

File: codes/elixir/chapter04/answer_A16.exs
defmodule DynamicProgramming do
  def main do
    n = String.trim(IO.gets("")) |> String.to_integer
    a = IO.read(:line) |> String.split() |> Enum.map(&String.to_integer/1)
    b = IO.read(:line) |> String.split() |> Enum.map(&String.to_integer/1)

    dp = Enum.reduce(1..n, %{}, fn i, acc ->
      case i do
        1 -> Map.put(acc, i, 0)
        2 -> Map.put(acc, i, Enum.at(a, 0))
        _ ->
          dp_i = [acc[i - 1] + Enum.at(a, i - 2), acc[i - 2] + Enum.at(b, i - 3)] |> Enum.min()
          Map.put(acc, i, dp_i)
      end
    end)

    IO.puts dp[n]
  end
end

DynamicProgramming.main()
